python/rtypes/dataframe/trigger_manager.py

```diff
@@ -0,0 +1,72 @@
+
+from enums import ObjectType
+import os
+
+
+class TriggerManager(object):
+    """ TriggerManager regulates the triggers in the dataframe:
+        - giving a space for triggers to live (See other managers for an approach to this)
+        - allowing users to add triggers into the manager
+        - allowing users to delete triggers into the manager
+        - preventing triggers of the same time and action from existing
+        - allow the user to add triggers with a specified priority and have that priority take affect with any overlapping triggers
+        - allow a way to access triggers saved in the manager
+    """
+
+    def __init__(self):
+        
+        # group key (always string fullname of type) to members of the group.
+        # Ex: group key: "Customer" -- > before_insert_customer, after_delete_customer, before_read_customer
+        self.trigger_group_members = dict()
+
+    #################################################
+    ### API Methods #################################
+    #################################################
+
+    def add_trigger(self, trigger_obj):
+        # can add a single Trigger class into the trigger_manager into the dataframe
+
+        # Question: Like the type_manager's add_type method, does this method need to return a set of pairs that have been added?
+
+        self.__add_type(trigger_obj)
+
+    def add_types(self, trigger_objs):
+        # can add multiple Trigger classes into the trigger_manager into the 
+        for obj in trigger_objs:
+            self.__add_type(obj)
+
+    def remove_trigger(self, trigger_obj):
+        # can remove a Trigger class from the trigger_manager into the dataframe
+        pass
+
+    #################################################
+    ### Private Methods #############################
+    #################################################
+
+    def __add_type(self, trigger_obj):
+        """ The manager adds the trigger to a group based on the object type that it is associated with
+            - If that associated object type already has a trigger, the manager checks if their is a trigger with the same action and time
+                - If it does, then it place the trigger in the group accordingly
+                - If it doesn't then it will place the trigger in the list
+            - groups are created in dictionaries in the following format: {"associated_pcc_type": {"action+time": trigger_object}}
+        """
+        # Ex: {"Customer": {"beforeinsert": [before_insert_Customer]}}
+
+        if trigger_obj.type in self.trigger_group_members.keys():
```

Comment:
`trigger_obj.type in self.trigger_group_members:`
Don't use keys()
makes the lookup slower (O(n) instead of O(1))
Read hash tables in CLR and compare lookup in hash tables to lookup in arrays to understand the difference.